README.md
 **This work has been accepted in Machine Learning and Applications as of December 17 2023. Please cite the paper as follows until further update**

Brittany Ho, Ta’Rhonda Mayberry, Khanh Linh Nguyen, Manohar Dhulipala, Vivek Krishnamani Pallipuram,
ChatReview: A ChatGPT-enabled natural language processing framework to study domain-specific user reviews,
Machine Learning with Applications, Volume 15, 2024, 100522, ISSN 2666-8270, https://doi.org/10.1016/j.mlwa.2023.100522.


# Abstract

We present ChatReview, a ChatGPT-enabled natural language processing framework that effectively studies domain-specific user
reviews to offer relevant and personalized search results at multiple levels of granularity. The framework accomplishes this task
using four phases including data collection, tokenization, query construction, and response generation. The data collection phase
involves gathering domain-specific user reviews from public and private repositories. In the tokenization phase, ChatReview applies
sentiment analysis to extract keywords and categorize them into various sentiment classes. This process creates a token repository
that best describes the user sentiments for a given user-review data. In the query construction phase, the framework uses the
token repository and domain knowledge to construct three types of ChatGPT prompts including explicit, implicit, and creative.
In the response generation phase, ChatReview pipelines these prompts into ChatGPT to generate search results at varying levels
of granularity.

# Chat Review Analyzer

This Python script analyzes reviews, visualizes emotional content using NLP (Natural Language Processing) techniques, and generates recommendations using ChatGPT API.

## Prerequisites

Before running the script, ensure you have Python installed on your system. You will also need to install the required packages listed in requirements.txt. You can install these packages using the following command:

Copy and run code in your terminal inside src folder:

<code>pip3 install -r requirements.txt </code>

## Configuration

To use this script, you will need to modify the following configurations based on your needs:

- File Paths: Update the paths for input and output files within the script.
- Output File Names: Customize the names of the output files generated by the script.
- TARGET_EMOTIONS: Adjust this variable to focus on specific emotions relevant to your analysis.

Specifically, modify the following fields (at the top of the script):
```python
openai.api_key = ""
REVIEWS_FILE_PATH = '../data/Hospitality_Data/HospitalityReviews.txt'
CATEGORIES_QUESTIONS_FILEPATH = '../data/Hospitality_Data/hospitallity_explicit.csv'
# Education: ["negative", "anger", "sadness"]
# Restaurant: ["negative", "anger", "disgust"]
# Hospitality: ["negative", "anger", "fear", "disgust"]
TARGET_EMOTIONS = ["negative", "anger", "fear", "disgust"]
WORD_CLOUD_PNG ="../OUTPUT/wordcloud_hospitality.png"
OUTPUT_FILEPATH = "../OUTPUT/output_hospitality.txt"
```



# OpenAI API Key

The script uses OpenAI's API for some of its functionality. You will need to obtain your own OpenAI API key. To do this, follow these steps:

- Create an account on the OpenAI website.
- Navigate to the API section and follow the instructions to generate an API key.
- Replace the existing API key in the script with your key.
Note: Keep your API key confidential to prevent unauthorized usage.

# Usage

Run the script using the following command:

Copy code
<code>python3 ChatReview.py</code>

Ensure that you have set up all the necessary configurations and have your OpenAI API key in place before executing the script.


History.txt
**Major Updates**

December 28, 2023: Upgraded transformers to version 4.36. Earlier versions have been reported to have deserialization issues.


src/QueryContruction_Education.py
#!/usr/bin/env python
# coding: utf-8

# In[3]:


from nrclex import NRCLex
import pprint

pp = pprint.PrettyPrinter(indent=4)

with open('./Education_Data/evals.txt', 'r', encoding='utf-8') as f:
    reviews = f.readlines()

reviews = ';'.join(reviews)
text_object = NRCLex(reviews)

original_dict = text_object.affect_dict
result_dict = {}

for word, emotions in original_dict.items():
    for emotion in emotions:
        if emotion not in result_dict:
            result_dict[emotion] = []
        result_dict[emotion].append(word)


# In[ ]:


from collections import Counter
from wordcloud import WordCloud
import matplotlib.pyplot as plt
import re
from nltk.corpus import stopwords
import operator
from pprint import pprint

# Read the text file
with open('./Education_Data/evals.txt', 'r', encoding='utf-8') as f:
    dataset = f.read()

# Clean the dataset
cleaned_dataset = re.sub(r'[^\w\s]', '', dataset)  # Remove punctuation
cleaned_dataset = re.sub(r'\d+', '', cleaned_dataset)  # Remove numbers
cleaned_dataset = cleaned_dataset.lower()  # Convert to lowercase

# Remove stop words
stop_words = set(stopwords.words('english'))
words = cleaned_dataset.split()
filtered_words = [word for word in words if word not in stop_words]

# Join the filtered words back into a string
cleaned_dataset = ' '.join(filtered_words)

# Print the cleaned dataset
# print(cleaned_dataset)

counter = {}

# Split the string into words
words = cleaned_dataset.split()

# Count the occurrences of each word
for word in words:
    if word not in counter:
        counter[word] = 0
    counter[word] += 1

pprint(counter) #dict


# In[ ]:


combined_list = []
def WordsByEmotions(emotionStrs):
    # Dictionary to store individual word frequencies for each emotion
    individual_word_frequencies = {emotion: {} for emotion in emotionStrs}

    # Create a dictionary to store combined frequencies of words across emotions
    combined_word_frequencies = {}

    # Iterate over all given emotions
    for emotionStr in emotionStrs:
        emotion_words = result_dict.get(emotionStr, [])
        
        for word in emotion_words:
            # Check if the word exists in the counter dictionary
            if word in counter:
                # For individual word frequencies per emotion
                if word not in individual_word_frequencies[emotionStr]:
                    individual_word_frequencies[emotionStr][word] = 0
                individual_word_frequencies[emotionStr][word] += counter[word]

                # Add the frequency of the word to the combined_word_frequencies dictionary
                if word not in combined_word_frequencies:
                    combined_word_frequencies[word] = 0
                combined_word_frequencies[word] += counter[word]

    # Print individual word frequencies for each emotion
    for emotion, word_freqs in individual_word_frequencies.items():
        sorted_word_freqs = sorted(word_freqs.items(), key=operator.itemgetter(1), reverse=True)
        print(f"Word Frequencies for {emotion}:")
        for word, freq in sorted_word_freqs:
            print(f"{word}: {freq}")
        print("\n")

    # Print sorted combined word frequencies
    sorted_combined_word_freqs = sorted(combined_word_frequencies.items(), key=operator.itemgetter(1), reverse=True)
    print("Combined Word Frequencies:")
    for word, freq in sorted_combined_word_freqs:
        combined_list.append(word)
        print(f"{word}: {freq}")
    print("\n")

    # Create the word cloud from the combined frequencies
    wordcloud = WordCloud(width=600, height=300, background_color='white', max_font_size=60)
    wordcloud.generate_from_frequencies(combined_word_frequencies)
    
    # Display the word cloud
    plt.figure(figsize=(10, 5))
    plt.imshow(wordcloud, interpolation='bilinear')
    title = 'Words Associated with ' + ", ".join(emotionStrs)
    plt.title(title)
    plt.axis('off')
    plt.savefig("wordcloud.png", dpi=300, bbox_inches='tight')
    plt.show()


# Generate word cloud for multiple emotions and print sorted counts
WordsByEmotions(["negative", "anger", "sadness"])
# WordsByEmotion("negative")
# WordsByEmotion("anger")
# WordsByEmotion("sadness")


# In[ ]:


import pandas as pd

# Read the csv file
df = pd.read_csv('./Education_Data/paper_user_education_creative.csv')

# Get the 'Categories' column and convert it to a list
categories = df['Category'].tolist()
keywords = combined_list
print(combined_list)


# In[ ]:


from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

# Initialize the model (it will download the parameters the first time)
model = SentenceTransformer('all-MiniLM-L6-v2')

# Create vectors for categories and keywords
category_vectors = {category: model.encode(category) for category in categories}
keyword_vectors = {keyword: model.encode(keyword) for keyword in keywords}

# Create a dictionary to hold the grouped keywords for each category
grouped_keywords = {category: [] for category in categories}

for category, category_vector in category_vectors.items():
    for keyword, keyword_vector in keyword_vectors.items():
        similarity = cosine_similarity([category_vector], [keyword_vector])[0][0]
        if similarity > 0.25:  # you may want to adjust this threshold
            grouped_keywords[category].append(keyword)

# Print grouped keywords
for category, keywords in grouped_keywords.items():
    print(f"Category:", f"{category}\nKeywords: {keywords}\n")


# In[ ]:


# Convert grouped keywords into a list of tuples (category, keyword count)
keyword_counts = [(category, len(keywords)) for category, keywords in grouped_keywords.items()]

# Sort by keyword count in descending order
keyword_counts.sort(key=lambda x: x[1], reverse=True)

# Print the top 3 categories
cat = []
for category, count in keyword_counts[:26]:
    print(f"Category: {category}\nKeyword count: {count}\n")
    cat.append(category)
    
print(cat)
    


# ### Organization (for each of the categories, pre-assigned questions)
# --> 26 questions
# Next meeting: run Manohar's data --> word clouds --> categories --> questions to GPT --> Result from ChatGPT
# 
# Present: in action, randomly select 50/150/180 reviews from database (after expanded)
# 2 demonstrations: 1/ How the code works 2/general public/audience presentation (storyline/powerpoint)

# In[ ]:


# Create a dictionary for easier lookup
question_dict = pd.Series(df.Question.values,index=df.Category).to_dict()

# Print the top 3 categories with their corresponding questions
for category, count in keyword_counts[:3]:
    print(f"Category: {category}\nKeyword count: {count}\nQuestion: {question_dict[category]}\n")


# In[ ]:


import openai
openai.api_key = "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"


# In[ ]:


import pandas as pd
import openai
from termcolor import colored
# Get the top 3 categories with their corresponding questions
questions = [question_dict[category] for category, _ in keyword_counts[:3]]

# Use each question as input to the GPT-3.5-turbo model
for question in questions:
    print(f"Question: {question}\n")
    
    output = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "user", "content": question}
        ],
        temperature=0,
        stream=True
    )
    
    # Iterate through the chunks and print the content
    for chunk in output:
        if 'content' in chunk['choices'][0]['delta']:
            print(colored(chunk['choices'][0]['delta']['content'], 'blue'), end = "")

    print("\n")


# In[ ]:






src/QueryContruction_Restaurant.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


from nrclex import NRCLex
import pprint

pp = pprint.PrettyPrinter(indent=4)

with open('./Restaurant_Data/OGI_Restaurant_40.txt', 'r', encoding='utf-8') as f:
    reviews = f.readlines()

reviews = ';'.join(reviews)
text_object = NRCLex(reviews)

original_dict = text_object.affect_dict
result_dict = {}

for word, emotions in original_dict.items():
    for emotion in emotions:
        if emotion not in result_dict:
            result_dict[emotion] = []
        result_dict[emotion].append(word)


# In[2]:


from collections import Counter
from wordcloud import WordCloud
import matplotlib.pyplot as plt
import re
from nltk.corpus import stopwords
import operator
from pprint import pprint

# Read the text file
with open('./Restaurant_Data/OGI_Restaurant_40.txt', 'r', encoding='utf-8') as f:
    dataset = f.read()

# Clean the dataset
cleaned_dataset = re.sub(r'[^\w\s]', '', dataset)  # Remove punctuation
cleaned_dataset = re.sub(r'\d+', '', cleaned_dataset)  # Remove numbers
cleaned_dataset = cleaned_dataset.lower()  # Convert to lowercase

# Remove stop words
stop_words = set(stopwords.words('english'))
words = cleaned_dataset.split()
filtered_words = [word for word in words if word not in stop_words]

# Join the filtered words back into a string
cleaned_dataset = ' '.join(filtered_words)

# Print the cleaned dataset
# print(cleaned_dataset)

counter = {}

# Split the string into words
words = cleaned_dataset.split()

# Count the occurrences of each word
for word in words:
    if word not in counter:
        counter[word] = 0
    counter[word] += 1

pprint(counter) #dict


# In[3]:


combined_list = []
def WordsByEmotions(emotionStrs):
    # Dictionary to store individual word frequencies for each emotion
    individual_word_frequencies = {emotion: {} for emotion in emotionStrs}

    # Create a dictionary to store combined frequencies of words across emotions
    combined_word_frequencies = {}

    # Iterate over all given emotions
    for emotionStr in emotionStrs:
        emotion_words = result_dict.get(emotionStr, [])
        
        for word in emotion_words:
            # Check if the word exists in the counter dictionary
            if word in counter:
                # For individual word frequencies per emotion
                if word not in individual_word_frequencies[emotionStr]:
                    individual_word_frequencies[emotionStr][word] = 0
                individual_word_frequencies[emotionStr][word] += counter[word]

                # Add the frequency of the word to the combined_word_frequencies dictionary
                if word not in combined_word_frequencies:
                    combined_word_frequencies[word] = 0
                combined_word_frequencies[word] += counter[word]

    # Print individual word frequencies for each emotion
    for emotion, word_freqs in individual_word_frequencies.items():
        sorted_word_freqs = sorted(word_freqs.items(), key=operator.itemgetter(1), reverse=True)
        print(f"Word Frequencies for {emotion}:")
        for word, freq in sorted_word_freqs:
            print(f"{word}: {freq}")
        print("\n")

    # Print sorted combined word frequencies
    sorted_combined_word_freqs = sorted(combined_word_frequencies.items(), key=operator.itemgetter(1), reverse=True)
    print("Combined Word Frequencies:")
    for word, freq in sorted_combined_word_freqs:
        combined_list.append(word)
        print(f"{word}: {freq}")
    print("\n")

    # Create the word cloud from the combined frequencies
    wordcloud = WordCloud(width=600, height=300, background_color='white', max_font_size=60)
    wordcloud.generate_from_frequencies(combined_word_frequencies)
    
    # Display the word cloud
    plt.figure(figsize=(10, 5))
    plt.imshow(wordcloud, interpolation='bilinear')
    title = 'Words Associated with ' + ", ".join(emotionStrs)
    plt.title(title)
    plt.axis('off')
    plt.savefig("wordcloud.png", dpi=300, bbox_inches='tight')
    plt.show()


# Generate word cloud for multiple emotions and print sorted counts
WordsByEmotions(["negative", "anger", "disgust"])
# WordsByEmotion("negative")
# WordsByEmotion("anger")
# WordsByEmotion("sadness")


# In[4]:


import pandas as pd

# Read the csv file
df = pd.read_csv('./Restaurant_Data/paper_user_restaurant_explicit.csv')

# Get the 'Categories' column and convert it to a list
categories = df['Category'].tolist()
keywords = combined_list
print(combined_list)


# In[5]:


from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

# Initialize the BERT model (it will download the parameters the first time)
model = SentenceTransformer('all-MiniLM-L6-v2')

# Create vectors for categories and keywords
category_vectors = {category: model.encode(category) for category in categories}
keyword_vectors = {keyword: model.encode(keyword) for keyword in keywords}

# Create a dictionary to hold the grouped keywords for each category
grouped_keywords = {category: [] for category in categories}

for category, category_vector in category_vectors.items():
    for keyword, keyword_vector in keyword_vectors.items():
        similarity = cosine_similarity([category_vector], [keyword_vector])[0][0]
        if similarity > 0.25:  # we may want to adjust this threshold
            grouped_keywords[category].append(keyword)

# Print grouped keywords
for category, keywords in grouped_keywords.items():
    print(f"Category:", f"{category}\nKeywords: {keywords}\n")


# In[6]:


# Convert grouped keywords into a list of tuples (category, keyword count)
keyword_counts = [(category, len(keywords)) for category, keywords in grouped_keywords.items()]

# Sort by keyword count in descending order
keyword_counts.sort(key=lambda x: x[1], reverse=True)

# Print the top 3 categories
cat = []
for category, count in keyword_counts[:10]:
    print(f"Category: {category}\nKeyword count: {count}\n")
    cat.append(category)
    
print(cat)
    


# ### Organization (for each of the categories, pre-assigned questions)
# --> 26 questions
# Next meeting: run Manohar's data --> word clouds --> categories --> questions to GPT --> Result from ChatGPT
# 
# Present: in action, randomly select 50/150/180 reviews from database (after expanded)
# 2 demonstrations: 1/ How the code works 2/general public/audience presentation (storyline/powerpoint)

# In[7]:


# Create a dictionary for easier lookup
question_dict = pd.Series(df.Question.values,index=df.Category).to_dict()

# Print the top 3 categories with their corresponding questions
for category, count in keyword_counts[:3]:
    print(f"Category: {category}\nKeyword count: {count}\nQuestion: {question_dict[category]}\n")


# In[8]:


import openai
openai.api_key = "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"


# In[9]:


import pandas as pd
import openai
from termcolor import colored
# Get the top 3 categories with their corresponding questions
questions = [question_dict[category] for category, _ in keyword_counts[:3]]

# Use each question as input to the GPT-3.5-turbo model
for question in questions:
    print(f"Question: {question}\n")
    
    output = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "user", "content": question}
        ],
        temperature=0,
        stream=True
    )
    
    # Iterate through the chunks and print the content
    for chunk in output:
        if 'content' in chunk['choices'][0]['delta']:
            print(colored(chunk['choices'][0]['delta']['content'], 'blue'), end = "")

    print("\n")



src/QueryContruction_Hospitality.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


from nrclex import NRCLex
from collections import Counter
from wordcloud import WordCloud
import matplotlib.pyplot as plt
import re
from nltk.corpus import stopwords
import operator
import pprint
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd
from termcolor import colored


# In[2]:


pp = pprint.PrettyPrinter(indent=4)

# Read the text file
with open('./Hospitality_Data/HospitalityReviews.txt', 'r', encoding='utf-8') as f:
    reviews = f.readlines()

reviews = ';'.join(reviews)
text_object = NRCLex(reviews)

original_dict = text_object.affect_dict
result_dict = {}

for word, emotions in original_dict.items():
    for emotion in emotions:
        if emotion not in result_dict:
            result_dict[emotion] = []
        result_dict[emotion].append(word)


# In[3]:


# Read the text file
with open('./Hospitality_Data/HospitalityReviews.txt', 'r', encoding='utf-8') as f:
    dataset = f.read()
    
# Clean the dataset
cleaned_dataset = re.sub(r'[^\w\s]', '', dataset)  # Remove punctuation
cleaned_dataset = re.sub(r'\d+', '', cleaned_dataset)  # Remove numbers
cleaned_dataset = cleaned_dataset.lower()  # Convert to lowercase

# Remove stop words
stop_words = set(stopwords.words('english'))
words = cleaned_dataset.split()
filtered_words = [word for word in words if word not in stop_words]

# Join the filtered words back into a string
cleaned_dataset = ' '.join(filtered_words)

counter = {}

# Split the string into words
words = cleaned_dataset.split()

# Count the occurrences of each word
for word in words:
    if word not in counter:
        counter[word] = 0
    counter[word] += 1

pp.pprint(counter) #dict


# In[4]:


combined_list = []
def WordsByEmotions(emotionStrs):
    # Dictionary to store individual word frequencies for each emotion
    individual_word_frequencies = {emotion: {} for emotion in emotionStrs}

    # Create a dictionary to store combined frequencies of words across emotions
    combined_word_frequencies = {}

    # Iterate over all given emotions
    for emotionStr in emotionStrs:
        emotion_words = result_dict.get(emotionStr, [])
        
        for word in emotion_words:
            # Check if the word exists in the counter dictionary
            if word in counter:
                # For individual word frequencies per emotion
                if word not in individual_word_frequencies[emotionStr]:
                    individual_word_frequencies[emotionStr][word] = 0
                individual_word_frequencies[emotionStr][word] += counter[word]

                # Add the frequency of the word to the combined_word_frequencies dictionary
                if word not in combined_word_frequencies:
                    combined_word_frequencies[word] = 0
                combined_word_frequencies[word] += counter[word]

    # Print individual word frequencies for each emotion
    for emotion, word_freqs in individual_word_frequencies.items():
        sorted_word_freqs = sorted(word_freqs.items(), key=operator.itemgetter(1), reverse=True)
        print(f"Word Frequencies for {emotion}:")
        for word, freq in sorted_word_freqs:
            print(f"{word}: {freq}")
        print("\n")

    # Print sorted combined word frequencies
    sorted_combined_word_freqs = sorted(combined_word_frequencies.items(), key=operator.itemgetter(1), reverse=True)
    print("Combined Word Frequencies:")
    for word, freq in sorted_combined_word_freqs:
        combined_list.append(word)
        print(f"{word}: {freq}")
    print("\n")

    # Create the word cloud from the combined frequencies
    wordcloud = WordCloud(width=600, height=300, background_color='white', max_font_size=60)
    wordcloud.generate_from_frequencies(combined_word_frequencies)
    
    # Display the word cloud
    plt.figure(figsize=(10, 5))
    plt.imshow(wordcloud, interpolation='bilinear')
    title = 'Words Associated with ' + ", ".join(emotionStrs)
    plt.title(title)
    plt.axis('off')
    plt.savefig("wordcloud.png", dpi=300, bbox_inches='tight')
    plt.show()


# Generate word cloud for multiple emotions and print sorted counts
WordsByEmotions(["negative", "anger", "fear", "disgust"])
# WordsByEmotion("negative")


# In[5]:


# Read the csv file
df = pd.read_csv('./Hospitality_Data/hospitallity_creative.csv')

# Get the 'Categories' column and convert it to a list
categories = df['Category'].tolist()
keywords = combined_list
print(combined_list)


# In[6]:


# Initialize the BERT model (it will download the parameters the first time)
model = SentenceTransformer('all-MiniLM-L6-v2')

# Create vectors for categories and keywords
category_vectors = {category: model.encode(category) for category in categories}
keyword_vectors = {keyword: model.encode(keyword) for keyword in keywords}

# Create a dictionary to hold the grouped keywords for each category
grouped_keywords = {category: [] for category in categories}

for category, category_vector in category_vectors.items():
    for keyword, keyword_vector in keyword_vectors.items():
        similarity = cosine_similarity([category_vector], [keyword_vector])[0][0]
        if similarity > 0.25:  # We may want to adjust this threshold
            grouped_keywords[category].append(keyword)

# Print grouped keywords
for category, keywords in grouped_keywords.items():
    print(f"Category:", f"{category}\nKeywords: {keywords}\n")


# In[7]:


# Convert grouped keywords into a list of tuples (category, keyword count)
keyword_counts = [(category, len(keywords)) for category, keywords in grouped_keywords.items()]

# Sort by keyword count in descending order
keyword_counts.sort(key=lambda x: x[1], reverse=True)

# Print the top 3 categories
cat = []
for category, count in keyword_counts[:10]:
    print(f"Category: {category}\nKeyword count: {count}\n")
    cat.append(category)
    
print(cat)
    


# ### Organization (for each of the categories, pre-assigned questions)
# --> 26 questions
# Next meeting: run Manohar's data --> word clouds --> categories --> questions to GPT --> Result from ChatGPT
# 
# Present: in action, randomly select 50/150/180 reviews from database (after expanded)
# 2 demonstrations: 1/ How the code works 2/general public/audience presentation (storyline/powerpoint)

# In[8]:


# Create a dictionary for easier lookup
question_dict = pd.Series(df.Question.values,index=df.Category).to_dict()

# Print the top 3 categories with their corresponding questions
for category, count in keyword_counts[:3]:
    print(f"Category: {category}\nKeyword count: {count}\nQuestion: {question_dict[category]}\n")


# In[9]:


import openai
openai.api_key = "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"


# In[10]:


# Get the top 3 categories with their corresponding questions
questions = [question_dict[category] for category, _ in keyword_counts[:3]]

# Use each question as input to the GPT-3.5-turbo model
for question in questions:
    print(f"Question: {question}\n")
    
    output = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "user", "content": question}
        ],
        temperature=0,
        stream=True
    )
    
    # Iterate through the chunks and print the content
    for chunk in output:
        if 'content' in chunk['choices'][0]['delta']:
            print(colored(chunk['choices'][0]['delta']['content'], 'blue'), end = "")

    print("\n")


# In[ ]:






src/ChatReview.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


from nrclex import NRCLex
import pprint
from collections import Counter
from wordcloud import WordCloud
import matplotlib.pyplot as plt
import re
from nltk.corpus import stopwords
import operator
import pandas as pd
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
from termcolor import colored
import openai

openai.api_key = ""
REVIEWS_FILE_PATH = '../data/Hospitality_Data/HospitalityReviews.txt'
CATEGORIES_QUESTIONS_FILEPATH = '../data/Hospitality_Data/hospitallity_explicit.csv'
# Education: ["negative", "anger", "sadness"]
# Restaurant: ["negative", "anger", "disgust"]
# Hospitality: ["negative", "anger", "fear", "disgust"]
TARGET_EMOTIONS = ["negative", "anger", "fear", "disgust"]
WORD_CLOUD_PNG ="../OUTPUT/wordcloud_hospitality.png"
OUTPUT_FILEPATH = "../OUTPUT/output_hospitality.txt"


# In[2]:


# Function to read and clean dataset
def read_and_clean_dataset(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        dataset = f.read()

    # Remove punctuation, numbers, convert to lowercase
    dataset = re.sub(r'[^\w\s]', '', dataset)
    dataset = re.sub(r'\d+', '', dataset)
    dataset = dataset.lower()

    # Remove stop words
    stop_words = set(stopwords.words('english'))
    words = dataset.split()
    filtered_words = [word for word in words if word not in stop_words]
    return ' '.join(filtered_words)


# In[3]:


# Function to count word frequencies
def count_word_frequencies(dataset):
    counter = Counter(dataset.split())
    return counter


# In[4]:


# Function to analyze emotions in text
def analyze_emotions(dataset):
    text_object = NRCLex(dataset)
    return text_object.affect_dict


# In[5]:


# Function to generate word cloud
def generate_word_cloud(frequencies, file_name=WORD_CLOUD_PNG):
    wordcloud = WordCloud(width=600, height=300, background_color='white', max_font_size=60)
    wordcloud.generate_from_frequencies(frequencies)
    
    plt.figure(figsize=(10, 5))
    plt.imshow(wordcloud, interpolation='bilinear')
    plt.title('Words Associated with ' + ", ".join(TARGET_EMOTIONS))
    plt.axis('off')
    plt.savefig(file_name, dpi=300, bbox_inches='tight')
    plt.show()
    


# In[6]:


def words_with_emotions(emotion_dict, target_emotions):
    combined_word_frequencies = {}

    for emotion in target_emotions:
        words_with_emotions = [word for word, emotions in emotion_dict.items() if emotion in emotions]
        combined_word_frequencies[emotion] = words_with_emotions
    
    return combined_word_frequencies 


# In[7]:


# Function to process words by emotions
def process_words_by_emotions(combined_word_frequencies, word_frequencies):
    unique_words = set(word for category in combined_word_frequencies.values() for word in category)
    # Associating each unique word with its count
    word_frequencies = {word: word_frequencies[word] for word in unique_words}
    return word_frequencies


# In[8]:


def main():
    with open(OUTPUT_FILEPATH, 'w') as file:
        pp = pprint.PrettyPrinter(indent=4, stream=file)

        # Read and clean dataset
        cleaned_dataset = read_and_clean_dataset(REVIEWS_FILE_PATH)

        # Analyze emotions in the dataset
        emotion_dict = analyze_emotions(cleaned_dataset) # return a dict
        print("Emotion Dictionary (Words and their associated emotions):", file=file)
        pp.pprint(emotion_dict)  # Debug print

        # Count word frequencies
        word_frequencies = count_word_frequencies(cleaned_dataset) # return a collections.Counter
    #     print("Word Frequencies:")
    #     pp.pprint(word_frequencies)  # Debug print

        combined_word_freqs = words_with_emotions(emotion_dict, TARGET_EMOTIONS)
        print("\nTarget emotions and their associated keywords (target keywords):", file=file)
        pp.pprint(combined_word_freqs)
        r_emo = process_words_by_emotions(combined_word_freqs, word_frequencies)
        print("\nFrequency of Target Keywords: ", file=file)
        pp.pprint(r_emo)  # Debug print

        # Generate word cloud for multiple emotions
        if combined_word_freqs:  # Check if combined_word_freqs is not empty
            generate_word_cloud(r_emo)
            print("\n.....Generated Word Cloud: ", WORD_CLOUD_PNG, file=file)
        else:
            print("No words found for the specified emotions.", file=file)  # Debug message

        # Read the csv file
        df = pd.read_csv(CATEGORIES_QUESTIONS_FILEPATH)

        # Get the 'Category' column and convert it to a list
        categories = df['Category'].tolist()
    #     pp.pprint(categories)
        keywords = list(r_emo.keys())             
    #     print(keywords)

        # Initialize the model (it will download the parameters the first time)
        model = SentenceTransformer('all-MiniLM-L6-v2')

        # Create vectors for categories and keywords
        category_vectors = {category: model.encode(category) for category in categories}
        keyword_vectors = {keyword: model.encode(keyword) for keyword in keywords}

        # Create a dictionary to hold the grouped keywords for each category
        grouped_keywords = {category: [] for category in categories}

        for category, category_vector in category_vectors.items():
            for keyword, keyword_vector in keyword_vectors.items():
                similarity = cosine_similarity([category_vector], [keyword_vector])[0][0]
                if similarity > 0.25:  # you may want to adjust this threshold
                    grouped_keywords[category].append(keyword)

        # Convert grouped keywords into a list of tuples (category, keyword count)
        keyword_counts = [(category, len(keywords)) for category, keywords in grouped_keywords.items()]

        # Sort by keyword count in descending order
        keyword_counts.sort(key=lambda x: x[1], reverse=True)

        # Use one for loop to print Category, Keyword count, and Keywords
        print("\nCategories:", file=file)
        for category, count in keyword_counts:
            # Retrieving keywords for the current category
            keywords = grouped_keywords[category]
            print(f"Category: {category}\n\tKeyword count: {count}\n\tKeywords: {keywords}\n", file=file)

        # Create a dictionary for easier lookup
        question_dict = pd.Series(df.Question.values,index=df.Category).to_dict()

        # Print the top 3 categories with their corresponding questions
        cat = []
        print("\nTop 3 categories with their corresponding questions:", file=file)
        for category, count in keyword_counts[:3]:
            print(f"Category: {category}\n\tKeyword count: {count}\n\tQuestion: {question_dict[category]}\n", file=file)
            cat.append(category)
    #     pp.pprint(cat) # print top 3 categories

        # Get the top 3 categories with their corresponding questions
        questions = [question_dict[category] for category, _ in keyword_counts[:3]]

        print("\nGPT Generated Recommendations based on Questions:", file = file)
        # Use each question as input to the GPT-3.5-turbo model
        for question in questions:
            print(f"Question: {question}\n", file=file)

            output = openai.ChatCompletion.create(
                model="gpt-3.5-turbo",
                messages=[
                    {"role": "user", "content": question}
                ],
                temperature=0,
                stream=True
            )

            # Iterate through the chunks and print the content
            for chunk in output:
                if 'content' in chunk['choices'][0]['delta']:
#                     print(colored(chunk['choices'][0]['delta']['content'], 'blue'), end = "", file=file)
                     print(chunk['choices'][0]['delta']['content'], end = "", file = file)

            print("\n", file=file)


main()


# In[ ]:




